include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,615 @@
+   *     scalars[0] = value1;
+   *     scalars[1] = value2;
+   *   }
+   *
+   *   void unpack_values(const std::vector<double> &scalars)
+   *   {
+   *     Assert (scalars.size() ==2, ExcInternalError());
+   *     value1 = scalars[0];
+   *     value2 = scalars[1];
+   *   }
+   * };
+   * </code>
+   * This class can then be use as
+   * <code>
+   * distributed::QPManager<dim,MyQData> manager(fe,lhs,rhs);
+   * ...mark cells for refinement...
+   * manager.prepare_for_coarsening_and_refinement(tr,2); // there will be 2 values
+   * tr.execute_coarsening_and_refinement();
+   * ...initialize quadrature point data on new cells
+   * manager.interpolate(tr);
+   * </code>
+   * This approach can be extended to quadrature point data with Tensors of
+   * arbitrary order, although with a little bit more work in packing and
+   * unpacking of data.
+   *
+   * @note During mesh balancing children of a single cell can be attributed
+   * to different processors. As a corollary, their parent cells can exist on
+   * multiple processors.
+   *
+   * @note The functionality provided by this class can alternatively be achieved
+   * with the parallel::distributed::SolutionTransfer . However, that would require
+   * the following steps: (i) create an auxiliary DoFHandler with a (discontinuous Galerkin)
+   * FiniteElement which has enough components to represent all data stored at the
+   * quadrature points data. (ii) project the data to the FiniteElement space
+   * and thereby store results in vectors; (iii) use the parallel::distributed::SolutionTransfer
+   * to project those vectors to the new mesh; (iv) and finally project the
+   * data back to the quadrature points on the new mesh via FEValues class.
+   * The QPManager class aims at simplifying the whole process only by requiring
+   * that the quadrature point data class is derived from the QPData
+   * and thereby implements functions for packing and unpacking its data into a
+   * vector. Everything else will be done automatically.
+   *
+   * @author Denis Davydov, Jean-Paul Pelteret, 2016
+   */
+  template<int dim, typename DATA>
+  class QPManager : public QPManagerBase<typename parallel::distributed::Triangulation<dim>::cell_iterator, DATA>
+  {
+  public:
+    /**
+     * A typedef for a cell.
+     */
+    typedef typename parallel::distributed::Triangulation<dim>::cell_iterator CELL;
+
+    /**
+     * Constructor which takes the FiniteElement @p projection_fe , the quadrature
+     * rule @p lhs_quadrature used to integrate its local mass matrix and
+     * finally the quadrature rule @p rhs_quadrature used to store DATA .
+     */
+    QPManager(const FiniteElement<dim> &projection_fe,
+              const Quadrature<dim> &lhs_quadrature,
+              const Quadrature<dim> &rhs_quadrature);
+
+    /**
+     * Initialize data at the cell @p cell .
+     */
+    template <typename T=DATA>
+    void initialize(const CELL &cell);
```

Comment:
By the way,
```
template <typename T=DATA>
    void initialize(const CELL &cell);
```
uses a defaulted template argument, which is C++11 only. It would certainly help if `T` was documented :-)